src/utils/process.rs

```diff
@@ -421,58 +437,105 @@ where
-                );
-                WithArgs {}
-            }
-            pub fn get() -> Option<Vec<String>> {
-                FAKE_ARGS.with(|a| {
-                    let old_value = a.replace_with(|old_value| match old_value {
-                        TlsState::Some(_) => TlsState::Invalid,
-                        TlsState::None => TlsState::None,
-                        TlsState::Invalid => TlsState::Invalid,
-                    });
-
-                    match old_value {
-                        TlsState::Some(args) => Some(args),
-                        TlsState::None => None,
-                        TlsState::Invalid => {
-                            panic!("test logic error (in get): wrong WithArgs scope?")
-                        }
-                    }
-                })
+        fn new<F>(args: &str, initial: F, from_: &str) -> Self
+        where
+            F: Fn(Vec<String>) -> TlsState<Vec<String>>,
+        {
+            let string_vec = args.split(' ').map(str::to_owned).collect();
+            if FAKE_ARGS.with(|a| a.replace(initial(string_vec))) != TlsState::None {
+                Self::error(from_);
             }
+            FakeParentArgs {}
         }
-        impl Drop for WithArgs {
-            fn drop(&mut self) {
-                // clears an invalid state
-                FAKE_ARGS.with(|a| a.replace(TlsState::None));
+        pub fn with(args: &[&str]) -> Self {
+            let with = TlsState::With(
+                0,
+                Rc::new(
+                    args.iter()
+                        .map(|a| a.split(' ').map(str::to_owned).collect())
+                        .collect(),
+                ),
+            );
+            if FAKE_ARGS.with(|a| a.replace(with)) != TlsState::None || args.is_empty() {
+                Self::error("with creation");
             }
+            FakeParentArgs {}
+        }
+        pub fn get() -> Option<Vec<String>> {
+            FAKE_ARGS.with(|a| {
+                let old_value = a.replace_with(|old_value| match old_value {
+                    TlsState::Once(_) => TlsState::Invalid,
+                    TlsState::Scope(args) => TlsState::Scope(args.clone()),
+                    TlsState::With(n, args) => TlsState::With(*n + 1, Rc::clone(args)),
+                    TlsState::None => TlsState::None,
+                    TlsState::Invalid => TlsState::Invalid,
+                });
+
+                match old_value {
+                    TlsState::Once(args) | TlsState::Scope(args) => Some(args),
+                    TlsState::With(n, args) if n < args.len() => Some(args[n].clone()),
+                    TlsState::None => None,
+                    TlsState::Invalid | TlsState::With(_, _) => Self::error("get"),
+                }
+            })
+        }
+        fn error(where_: &str) -> ! {
+            panic!(
+                "test logic error (in {}): wrong FakeParentArgs scope?",
+                where_
+            );
+        }
+    }
```

Comment:
These interior mutability pointer tricks are new to me, very nice! I'd heard of them while reading Blandy & Orendorff but hadn't tried to understand code using them before. Your comment above was sufficient though to explain what was going on, if taken slowly.